src/bst.py

```diff
@@ -123,6 +145,99 @@ def balance(self):
 
         return depth_left - depth_right
 
+    def _traverse(self, add, remove, size):
+        '''Traverse function
+            takes in other functions and a start_node'''
+        while size():
+            x = remove()
+            if x:
+                add(x.left, x.right)
+                yield x.value
+
+    def breadth_first_traversal(self):
+        '''perform a breadth first traversal, returns a list of
+           nodes in the graph
+        '''
+        q = Queue()
+        q.enqueue(self.root)
+
+        def add(a, b):
+            q.enqueue(a)
+            q.enqueue(b)
+
+        return self._traverse(add, q.dequeue, q.size)
+
+    def pre_order(self):
+        '''perform a breadth first traversal, returns a list of
+           nodes in the graph
+        '''
+        s = Stack()
+        s.push(self.root)
+
+        def add(a, b):
+            s.push(b)
+            s.push(a)
+        return self._traverse(add, s.pop, s.size)
+
+    def in_order(self):
+        '''Traverse function
+            takes in other functions and a start_node'''
+        if self.root:
+            return self.root.in_order()
+        else:
+            return []
+
+    def post_order(self):
+        '''Traverse function
+            takes in other functions and a start_node'''
+        if self.root:
+            return self.root.post_order()
+        else:
+            return []
+
+    def delete(self, target):
+        """a function that removes a child node"""
+        # initialization
+        current = self.root
+        pointer = None
+
+        # search
+        while current:
+            if target == current.value:
+                break
+            pointer = current
+            if current.value < target:
+                current = current.right
+            else:
+                current = current.left
+        if current is None:
+            return
+
+        # target has 2 children
+        if current.left is not None and current.right is not None:
+            y = current.left
+            pointer = current
+            while y.right:
+                pointer = y
+                y = y.right
+            current.value = y.value
+            current = y
+
+        # leaf and 1 child cases
+        if pointer is None:  # case where there is only one node
+            if current.left:
+                self.root = current.left
+            else:
+                self.root = current.right
+        if current.left:
+            tmp = current.left
+        else:
+            tmp = current.right
+        if current == pointer.right:
```

Comment:
This is throwing an `AttributeError` on many of my tests, including the cases where there is only one item in the tree and where there are two items, as well as some others. Make sure you are testing these always and forever.